File: src/main/rust/g0001_0100/s0008_string_to_integer_atoi/Solution.rs
// #Medium #Top_Interview_Questions #String #2024_09_04_Time_0_ms_(100.00%)_Space_2.1_MB_(82.19%)

impl Solution {
    pub fn my_atoi(str: String) -> i32 {
        let mut chars = str.chars().peekable();
        let mut i = 0;
        let mut negative_sign = false;
        let mut num: i32 = 0;

        // Skip leading whitespaces
        while let Some(&ch) = chars.peek() {
            if ch == ' ' {
                chars.next();
            } else {
                break;
            }
        }

        // Handle sign
        if let Some(&ch) = chars.peek() {
            if ch == '+' {
                chars.next();
            } else if ch == '-' {
                chars.next();
                negative_sign = true;
            }
        }

        // Convert characters to integer
        while let Some(&ch) = chars.peek() {
            if ch >= '0' && ch <= '9' {
                chars.next();
                let digit = ch.to_digit(10).unwrap() as i32;
                let digit = if negative_sign { -digit } else { digit };

                // Check for overflow/underflow
                if num > i32::MAX / 10 || (num == i32::MAX / 10 && digit > 7) {
                    return i32::MAX;
                }
                if num < i32::MIN / 10 || (num == i32::MIN / 10 && digit < -8) {
                    return i32::MIN;
                }

                num = num * 10 + digit;
            } else {
                break;
            }
        }

        num
    }
}

File: src/main/dart/g0001_0100/s0008_string_to_integer_atoi/Solution.dart
// #Medium #Top_Interview_Questions #String #2024_09_30_Time_349_ms_(85.92%)_Space_152.5_MB_(8.45%)

class Solution {
  static const digits = {
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "0": 0
  };
  static const MAX = 2147483647;
  static const MIN = -2147483648;

  int myAtoi(String s) {
    int res = 0;
    int sign = 1;
    int current = 0;
    // skip whitespaces
    while (current < s.length && s[current] == ' ') {
      current++;
    }
    //read sign
    if (current < s.length && (s[current] == '-' || s[current] == '+')) {
      sign = s[current++] == '-' ? -1 : 1;
    }
    // read digits
    while (current < s.length && digits.containsKey(s[current])) {
      int digit = digits[s[current++]]!;
      // check overflow
      if (sign == -1 && res < (MIN + digit) / 10) {
        return MIN;
      } else if (res > (MAX - digit) / 10) {
        return MAX;
      }
      res = res * 10 + sign * digit;
    }
    return res;
  }
}
